openraft/src/core/admin.rs

```diff
@@ -108,6 +136,33 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
             // non-blocking mode, do not know about the replication stat.
             let _ = tx.send(Ok(AddLearnerResponse { matched: None }));
         }
+
+        tracing::debug!(
+            "after add target node {} as learner {:?}",
+            target,
+            self.core.last_log_id
+        );
+    }
+
+    pub fn has_pending_membership_change(&self, members: &BTreeSet<NodeId>) -> bool {
+        if self.core.committed < Some(self.core.effective_membership.log_id) {
+            return false;
+        }
+
+        let curr = &self.core.effective_membership.membership;
+        let all_learners = curr.all_learners();
+        for new_node in members.difference(curr.all_members()) {
+            match all_learners.get(new_node) {
+                Some(_node) => {
+                    continue;
+                }
+                None => {
+                    // if new member not include in the learners, there is pending membership change
+                    return true;
+                }
+            }
+        }
+        false
```

Comment:
Why is it a `pending` state when there is a new member that is not in `learners`?

Reply:
The first condition is to ensure no new membership would be proposed when there is uncommitted membership.
I do not quite understand why the second part is there. 🤔

Reply:
> The first condition is to ensure no new membership would be proposed when there is uncommitted membership. I do not quite understand why the second part is there. 🤔

in `Raft::change_membership`,first call `add_learner` and then `change_membership`,so when `add_learner` add a new log, `change_membership` cannot process in previous condition.

so i refactor the `change_member` condition judge into two case in function `has_pending_membership_change `:

1. if or not has a biggger commit log.
2. if there is new member not in the learner set.

Reply:
It seems we should add learners and change members in one log entry.
Then there won't be such a problem.

Another issue is that we need to spawn replication tasks just after `append_payload_to_log()` and before `replicate_client_request()`.

What say you man?

Reply:
> It seems we should add learners and change members in one log entry. Then there won't be such a problem.

yes,i agree with this,i will refactor the source.

> Another issue is that we need to spawn replication tasks just after append_payload_to_log() and before replicate_client_request().

ok, assign a new issue to me.